Account for overhead in Cluster Scaler

adamnovak opened the pull request:
This will fix #2103. This is also relevant to https://github.com/DataBiosphere/toil/issues/4147#issuecomment-1179587561 where a node with memory size very close to the job's requirements was provisioned but not able to be used.

This changes the cluster scaler so that it preprocesses its node sizes to remove some resources for overhead (i.e. the kernel, other processes on the system, and Mesos/Kubernetes). Currently we reserve memory and disk.

To guess at how much memory might be used, in the absence of much real data, we use [a formula from Google Kubernetes Engine](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#memory_cpu) which describes how much memory GKE clusters keep for themselves, as a function of node memory size. We assume that Toil-managed Mesos and Kubernetes clusters use that much memory or less.

To guess at how much disk space might be used, we take 5% of the disk space (commonly a fraction reserved for superuser) and add 5 GiB. If this is more than the size of the disk on the node, we complain that the disk is weirdly small and assume all of it will be reserved (which may come back to haunt the autoscaler later if we ever try to use ~6 GB nodes).

None of this is user-configurable. **Unless you pass `--assumeZeroOverhead` in which case we turn it off.**

I've added some tests for made-up and real-world cases to constrain the overhead behavior to something sensible. Those pass, but I'm not sure if any live Toil autoscaler tests will now not scale up because they think they can't fit jobs where they will actually fit.

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * Toil's built-in autoscaler now guesses that some memory and disk space on nodes will not actually be available for jobs; pass `--assumeZeroOverhead` to revert to the old behavior (#2103) 

## Reviewer Checklist

<!-- To be kept in sync with docs/contributing/checklist.rst -->

 * [x] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [x] Read through the code changes. Make sure that it doesn't have:
    * [x] Addition of trailing whitespace.
    * [x] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [x] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [x] New functions or classes without informative docstrings.
    * [x] Changes to semantics not reflected in the relevant docstrings.
    * [x] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [x] New features without tests.
* [x] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [x] Finish the review with an overall description of your opinion.

## Merger Checklist

<!-- To be kept in sync with docs/contributing/checklist.rst -->

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

w-gao commented on src/toil/provisioners/clusterScaler.py:
```diff
@@ -416,6 +433,91 @@ def _round(self, number: float) -> int:
             # If we get here, something has gone wrong.
             raise RuntimeError(f"Could not round {number}")
 
+    def _reserve_overhead(self, full_node: Shape) -> Shape:
```
```
None of this is user-configurable.
```

Would we ever want to disable / overwrite the overhead? Some tests are not passing because the limit is set to a very small value, but I don't know if that will happen realistically.

adamnovak replied:
I think users might run into situations where we thing a job won't fit when it really will fit. It might make sense to give people a way to turn this back off, since it's possible that they already know a workflow works and we'd be guessing it wouldn't work. But I'm having trouble thinking of a good way to let people actually tune this.

w-gao commented on src/toil/test/provisioners/clusterScalerTest.py:
```diff
@@ -438,6 +450,121 @@ def testBetaInertia(self):
             scaler.smoothEstimate(c4_8xlarge_preemptable, 100)
         self.assertEqual(scaler.smoothEstimate(c4_8xlarge_preemptable, 100), 100)
 
+    def test_overhead_accounting_large(self):
```
great tests!

w-gao reviewed (commented):
This looks great to me! Some cluster scaler tests are not passing though:

```
INFO     toil.test:__init__.py:113 Setting up toil.test.provisioners.clusterScalerTest.ScalerThreadTest.testClusterScaling ...
WARNING  toil.provisioners.clusterScaler:clusterScaler.py:512 All 10.0 bB of disk on a node type are likely to be needed by the OS! The node probably cannot do any useful work!
WARNING  toil.provisioners.clusterScaler:clusterScaler.py:105 Couldn't fit job with requirements Shape(wallTime=9.671000000000003, memory=10, cores=10, disk=10, preemptable=False) into any nodes in the nodeTypes list.
```
https://github.com/DataBiosphere/toil/blob/0b754e37572a45fce6feb5b6ad46099a16f7034f/src/toil/test/provisioners/clusterScalerTest.py#L631

adamnovak commented on src/toil/provisioners/clusterScaler.py:
```diff
@@ -416,6 +433,91 @@ def _round(self, number: float) -> int:
+            # since the previous breakpoint, like a progressive income tax.
+            limit = min(breakpoint, memory_bytes)
+            reservation = fraction * (limit - accounted)
+            logger.debug('Reserve %sB of memory between %sB and %sB', bytes2human(reservation), bytes2human(accounted), bytes2human(limit))
+            reserved += reservation
+            accounted = limit
+            if accounted >= memory_bytes:
+                break
+        logger.debug('Reserved %sB/%sB memory for overhead', bytes2human(reserved), bytes2human(memory_bytes))
```
I kept ending up with like `123 Gi` when I really wanted `GiB`. But I suppose `Gi` is fine.

w-gao replied:
Ohh, that makes sense. I think the only problem was that something like `200 b` would end up being `200 bB`

adamnovak commented:
OK @w-gao I think I have addressed your comments. I *think( I've fixed the failing tests, by adapting the test cases to allow for the estimated overhead. I've also added an `--assumeZeroOverhead` flag that you can pass to skip shrinking the node sizes, and use the old autoscaling behavior.

w-gao reviewed (approved):
Looks great!
